bigquery_etl/shredder/delete.py

```diff
@@ -223,30 +224,76 @@ def create_job(client):
 
 def get_partition_expr(table):
     """Get the SQL expression to use for a table's partitioning field."""
-    for field in table.schema:
-        if field.name != table.time_partitioning.field:
-            continue
-        if field.field_type == "TIMESTAMP":
-            return f"DATE({field.name})"
-        return field.name
+    if table.range_partitioning:
+        return table.range_partitioning.field
+    if table.time_partitioning:
+        return next(
+            f"DATE({field.name})" if field.field_type == "TIMESTAMP" else field.name
+            for field in table.schema
+            if field.name == table.time_partitioning.field
+        )
+
+
+@dataclass
+class Partition:
+    """Return type for get_partition."""
+
+    condition: str
+    id: Optional[str] = None
 
 
-def list_partitions(client, target):
-    """List the partition ids and corresponding dates in a table."""
+def get_partition(table, partition_expr, end_date, id_=None) -> Optional[Partition]:
+    """Return a Partition for id_ unless it is a date on or after end_date."""
+    if id_ is None:
+        if table.time_partitioning:
+            return Partition(f"{partition_expr} < '{end_date}'")
+        return Partition("TRUE")
+    if id_ == "__NULL__":
```

Comment:
Where does this magic `"__NULL__"` value come from? Can this be a named constant with a descriptive name?

Reply:
`__NULL__` and `__UNPARTITIONED__` can be observed by creating a range partitioned table with values that are null and outside the configured range, and creating a time partitioned table with values that are null, and then listing the `partition_id`s with legacy SQL.

I became aware of this because it was causing errors because `__NULL__` isn't a valid date